# Ideal Gas Lab
My lovely groupmate already did the graphs so I am just getting some specific information I need from the graphs (plus practicing using this because I'll probably need it for the MAT proposal.

## Part 2

In [1]:
import pandas as pd

df = pd.read_csv("lab2-part2a.tsv", sep="\t")
print(df.head(10))

   Time(s)  T1(Deg C)  T2(Deg C)  P1(PSI)  P2(PSI)  Mass Flowrate(g/min)
0      0.0       25.6       25.8     23.8     24.4                   0.0
1      0.1       25.7       25.7     24.4     24.8                   0.0
2      0.2       25.7       25.8     24.9     24.4                   0.0
3      0.3       25.7       25.8     24.7     24.4                   0.0
4      0.4       25.8       25.8     24.4     25.0                   0.0
5      0.5       25.7       25.8     24.6     24.4                   0.0
6      0.6       25.7       25.8     24.3     24.4                   0.0
7      0.7       25.8       25.9     24.7     24.5                   0.0
8      0.8       25.9       25.9     24.9     24.7                   0.0
9      0.9       25.9       25.9     24.9     24.7                   0.0


### Get stabilized average T and P

From the data,
```
Ambient Temperature(C): 23.7
Ambient Pressure(in Hg): 29.8
```

From the Introductory Manual, 
```
1 in. Hg = 0.491 psi
```

I will do unit conversions on the lab report itself.

In [2]:
p_atm = 29.8 * 0.491
print(p_atm)

14.6318


In [3]:
# get relevant data
df2 = df[['T1(Deg C)', 'P1(PSI)']].copy()
print(df2.head(10))

   T1(Deg C)  P1(PSI)
0       25.6     23.8
1       25.7     24.4
2       25.7     24.9
3       25.7     24.7
4       25.8     24.4
5       25.7     24.6
6       25.7     24.3
7       25.8     24.7
8       25.9     24.9
9       25.9     24.9


In [4]:
# pressure
# change to absolute pressure
df2['P1(PSI)'] += p_atm

print(df2.head(10))

   T1(Deg C)  P1(PSI)
0       25.6  38.4318
1       25.7  39.0318
2       25.7  39.5318
3       25.7  39.3318
4       25.8  39.0318
5       25.7  39.2318
6       25.7  38.9318
7       25.8  39.3318
8       25.9  39.5318
9       25.9  39.5318


ABOVE, T1 = 25.6 ºC, P1 = 38.4318 PSI<br>
*if it is anything else you ran it too many times, rerun all previous cells including that one*

In [5]:
# try to find where temperature and pressure stabilize after rising
print(df2.iloc[295:340])

     T1(Deg C)  P1(PSI)
295       29.0  53.0318
296       29.0  52.9318
297       29.0  53.0318
298       29.1  53.3318
299       29.2  52.9318
300       29.2  53.2318
301       29.1  54.2318
302       29.3  53.6318
303       29.2  54.1318
304       29.2  54.1318
305       29.4  54.1318
306       29.2  54.6318
307       29.2  54.2318
308       29.2  54.7318
309       29.1  55.0318
310       29.1  54.9318
311       29.3  55.2318
312       29.2  54.6318
313       29.2  54.6318
314       29.0  54.4318
315       28.9  54.6318
316       28.9  54.2318
317       29.1  54.4318
318       28.8  54.2318
319       28.8  54.4318
320       28.9  54.6318
321       28.6  54.4318
322       28.6  54.4318
323       28.5  54.2318
324       28.5  54.6318
325       28.5  54.4318
326       28.3  54.2318
327       28.2  54.4318
328       28.2  54.6318
329       28.3  54.4318
330       28.2  54.2318
331       28.2  54.2318
332       28.0  54.6318
333       27.8  54.1318
334       27.8  54.2318
335       27.9  

In [6]:
# double check if still stable at the end
print(df2.tail(10))

     T1(Deg C)  P1(PSI)
816       26.1  54.1318
817       26.1  53.6318
818       26.1  54.2318
819       26.0  54.2318
820       26.0  54.2318
821       25.9  54.4318
822       26.0  54.4318
823       26.0  54.1318
824       26.2  53.8318
825       26.1  54.4318


In [7]:
# find average from 301 to 825 of T1 and P1
t1_stable, p1_stable = df2.loc[301:826].mean()
print(df2.loc[301:826].mean())
print(t1_stable)
print(p1_stable)

T1(Deg C)    26.595238
P1(PSI)      54.198657
dtype: float64
26.595238095238095
54.19865714285714


In [8]:
# find range of stable temperature and pressure at the beginning
print(df2.head(30))

    T1(Deg C)  P1(PSI)
0        25.6  38.4318
1        25.7  39.0318
2        25.7  39.5318
3        25.7  39.3318
4        25.8  39.0318
5        25.7  39.2318
6        25.7  38.9318
7        25.8  39.3318
8        25.9  39.5318
9        25.9  39.5318
10       25.7  39.2318
11       25.6  39.0318
12       25.6  39.3318
13       25.7  39.3318
14       25.7  39.6318
15       25.6  38.7318
16       25.7  39.2318
17       25.7  39.8318
18       25.5  39.8318
19       25.8  39.3318
20       25.8  39.3318
21       25.7  39.2318
22       25.7  39.5318
23       25.7  39.3318
24       25.7  39.2318
25       25.8  39.3318
26       25.8  39.3318
27       25.7  39.3318
28       25.7  39.2318
29       25.7  39.5318


In [9]:
print(df2.iloc[165:190])

     T1(Deg C)  P1(PSI)
165       25.7  39.2318
166       25.5  39.2318
167       25.8  39.3318
168       25.8  39.3318
169       25.9  39.3318
170       25.7  39.3318
171       25.7  38.7318
172       25.7  39.0318
173       25.7  41.3318
174       25.1  41.2318
175       26.0  41.2318
176       26.0  41.2318
177       26.1  41.0318
178       26.3  40.9318
179       26.3  40.9318
180       26.4  41.3318
181       26.5  41.0318
182       26.5  40.9318
183       26.5  41.2318
184       26.4  40.9318
185       26.4  41.0318
186       26.5  40.9318
187       26.4  41.0318
188       26.4  41.0318
189       26.3  40.7318


In [10]:
# find average from 0 to 172 of T1 and P1
t1_start, p1_start = df2.loc[0:173].mean()
print(df2.loc[0:173].mean())
print(t1_start)
print(p1_start)

T1(Deg C)    25.721264
P1(PSI)      39.301915
dtype: float64
25.721264367816094
39.30191494252873


I lied, let's do the unit conversions here
```
[K] = [ºC] + 273.15
1 psi = 6895 Pa
```

In [11]:
df3 = df2

df3['T1(Deg C)'] += 273.15
df3['P1(PSI)'] *= 6895

df3 = df3.rename(columns={'T1(Deg C)': 'T1(K)', 'P1(PSI)': 'P1(Pa)'})
print(df3.head(10))

    T1(K)      P1(Pa)
0  298.75  264987.261
1  298.85  269124.261
2  298.85  272571.761
3  298.85  271192.761
4  298.95  269124.261
5  298.85  270503.261
6  298.85  268434.761
7  298.95  271192.761
8  299.05  272571.761
9  299.05  272571.761


In [12]:
t1_avg, p1_avg = df3.loc[0:173].mean()
print(df3.loc[0:173].mean())
print(t1_avg)
print(p1_avg)

t2_avg, p2_avg = df3.loc[301:826].mean()
print(df3.loc[301:826].mean())
print(t2_avg)
print(p2_avg)

T1(K)        298.871264
P1(Pa)    270986.703529
dtype: float64
298.871264367816
270986.7035287356
T1(K)        299.745238
P1(Pa)    373699.741000
dtype: float64
299.7452380952381
373699.741


### Calculate Volume of the Left Tank, Initial mass

In [13]:
V_num = 9.20*10**(-3) * 0.2870 * 10**(3) * t1_avg * t2_avg
V_denom = p2_avg*t1_avg - p1_avg*t2_avg
V = V_num/V_denom
print(V)

0.0077653319046805565


In [14]:
m1 = (p1_avg*V)/(0.2870*10**3*t1_avg)
print(m1)

0.024532507898885413


### Uncertainties for Volume

In [26]:
# values
V = V
m_in = 9.20*10**(-3)
t_1 = t1_avg
t_2 = t2_avg
p_1 = p1_avg
p_2 = p2_avg
# uncertanties
d_m_in = 1/60 # 1 kg/min -> 1/60 kg/s
d_t_1 = 1 # 1ºC -> 1K
d_t_2 = 1
d_p_1 = 6895 # 1 psi -> 6895 Pa
d_p_2 = 6895

# unc for denom
term1 = (p_2*t_1)**2 * ( (d_p_2/p_2)**2 + (d_t_1/t_1)**2 )
term2 = (p_1*t_2)**2 * ( (d_p_1/p_1)**2 + (d_t_2/t_2)**2 )
d_denom = ( term1 + term2 )**0.5

print(d_denom)

# unc for V
term1 = (d_m_in/m_in)**2 + (d_t_1/t_1)**2 + (d_t_2/t_2)**2
term2 = (d_denom / (p_2*t_1 - p_1*t_2))**2
d_V = V * ( term1 + term2 )**0.5

print(d_V)


2954838.3560301107
0.014087830743856899


### Uncertainties for Initial Mass

In [27]:
# values
m1 = m1
p_1 = p1_avg
V = V
t_1 = t1_avg
# uncertanties
d_p_1 = 6895 # 1 psi -> 6895 Pa
d_V = d_V
d_t_1 = 1 # 1ºC -> 1K

# unc for m1
d_m1 = m1 * ( (d_p_1/p_1)**2 + (d_V/V)**2 + (d_t_1/t_1)**2 )**0.5
print(d_m1)

0.04451121987161814
